### PR TITLE
REP-4954 Form Encoding Support

### PR DESCRIPTION
**Pull request by wdschei**

### CODE REVIEW

File: repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala [hunk 1]
```diff
@@ -271,8 +245,46 @@ class HttpServletRequestWrapper(originalRequest: HttpServletRequest, inputStream
...
+        val updatedParameterMap = mutable.Map.empty[String, Array[String]]
+        formParameterMap match {
+          case Some(fpm) =>
+            updatedParameterMap ++= fpm
+          case None =>
+            // This is the only place the input stream has a chance of being replaced or manipulated directly by the wrapper.
+            inputStream synchronized {
+              // As per Servlet Spec 3.1 section 3.1.1, form parameters are only available until the input stream is read.
+              if (status == RequestBodyStatus.Available) {
+                if (!inputStream.markSupported) {
```
Comment (adrianjgeorge) [inline repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala]: We decided that we weren't going to reset the input stream ourselves. We were going to consume it if you asked for the parameters like everyone else does. If you want to be able to reset it, the implementer of the wrapper is responsible for wrapping the stream themselves before creating the wrapper.

**Comment by adrianjgeorge** [comment]
You didn't touch several of the parameter related methods which are still broke

File: repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala [hunk 2]
```diff
@@ -271,8 +245,46 @@ class HttpServletRequestWrapper(originalRequest: HttpServletRequest, inputStream
     *
     * @return the parameter map for this request
     */
-  override def getParameterMap: util.Map[String, Array[String]] =
-    parameterMap.map(_.asJava).getOrElse(super.getParameterMap)
+  override def getParameterMap: util.Map[String, Array[String]] = {
+    def retrieveParameterMap: util.Map[String, Array[String]] = {
+      // As per Servlet Spec 3.1 section 3.1.1, form parameters are only parsed under certain conditions.
+      if (getContentLength > 0 && "application/x-www-form-urlencoded".equalsIgnoreCase(getContentType) &&
```
Comment (adrianjgeorge) [inline repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala]: I hope there's a constant for this one

Comment (wdschei) [inline repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala]: Changed to a constant.

File: repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala (hunk 1 above)
Comment (wdschei) [inline repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala]: Updated.

**Comment by wdschei** [comment]
Updated the last two parameter related methods to use the return from the `getParameterMap()` method since the logic is already in there.

File: repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala [hunk 3]
```diff
@@ -271,8 +246,38 @@ class HttpServletRequestWrapper(originalRequest: HttpServletRequest, inputStream
...
+      // As per Servlet Spec 3.1 section 3.1.1, form parameters are only parsed under certain conditions.
+      if (getContentLength > 0 && APPLICATION_FORM_URLENCODED.equalsIgnoreCase(getContentType) &&
+        ("POST".equalsIgnoreCase(getMethod) || "PUT".equalsIgnoreCase(getMethod))) {
+        val updatedParameterMap = mutable.Map.empty[String, Array[String]]
+        formParameterMap match {
+          case Some(fpm) =>
+            updatedParameterMap ++= fpm
+          case None =>
+            // This is the only place the input stream has a chance of being replaced or manipulated directly by the wrapper.
+            inputStream synchronized {
```
Comment (adrianjgeorge) [inline repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala]: Probably unneeded at this point

File: repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala [hunk 4]
```diff
@@ -271,8 +246,38 @@ class HttpServletRequestWrapper(originalRequest: HttpServletRequest, inputStream
...
+              // As per Servlet Spec 3.1 section 3.1.1, form parameters are only available until the input stream is read.
+              if (status == RequestBodyStatus.Available) {
+                updatedParameterMap ++= parseParameterString(
+                  new String(RawInputStreamReader.instance.readFully(inputStream, getContentLength), StandardCharsets.UTF_8))
```
Comment (adrianjgeorge) [inline repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala]: Should probably mark the body as read

Comment (wdschei) [inline repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala]: By setting the `status`?

**Comment by adrianjgeorge** [comment]
Retest this please

**Pull request merged**